### ISSUE TITLE
Remove use of PyCrypto XOR "cipher" in Conch

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6386]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#6386|
|Type|enhancement|
|Created|2013-03-22 14:47:55Z|
|Branch|https://github.com/twisted/twisted/tree/use-hmac-6386-2|

There is often general rumbling about reducing the number of different crypto libraries Twisted depends on.  PyCrypto seems to be a popular target for removal, perhaps since Twisted's use of it is fairly small and largely confined to implementation details (not public interfaces).  The XOR "cipher" used by Conch for HMAC is low-hanging fruit - there are other ways to do XOR and other ways to do HMAC.

Also, PyCrypto itself at one point removed XOR and only re-instated it after we complained.  I doubt this situation will repeat itself, but perhaps it would be nice if we stopped depending on functionality PyCrypto does not actually want to provide.

<details><summary>Searchable metadata</summary>

```
trac-id__6386 6386
type__enhancement enhancement
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_use_hmac_6386_2 branches/use-hmac-6386-2
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__conch conch
keywords__ 
time__1363963675000000 1363963675000000
changetime__1370184312000000 1370184312000000
version__None None
owner__exarkun exarkun
cc__z3p
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

(In [37799]) Branching to 'use-hmac-6386'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun removed owner|
|-|-|

Changed it.  I don't really understand what's going on in `_getMAC`.  I wrote a comment but it's probably wrong.

[Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/use-hmac-6386).

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_4"></a>|@wallrj set owner to @wallrj<br>@wallrj set status to `assigned`|
|-|-|

Reviewing...

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_5"></a>|@wallrj set owner to @exarkun<br>@wallrj set status to `new`|
|-|-|

## Code Review

Hi Exarkun,

I'm no expert on conch or ssh mechanics, so here's a cosmetic code
review of your branch.

Notes:
 * Merges cleanly to trunk
 * Tests pass locally
 * Coverage seems complete.

Some minor points:
 1. One minor build failure
    1. https://buildbot.twistedmatrix.com/builders/twistedchecker/builds/546/steps/run-twistedchecker/logs/new%20twistedchecker%20errors
       ```
       ************* Module twisted.conch.test.test_transport
       W9012:1947,0: Expected 2 blank lines, found 3
       ```
 2. [branches/use-hmac-6386/twisted/conch/test/test_transport.py](branches/use-hmac-6386/twisted/conch/test/test_transport.py)
    1. test_getMAC
       1. typo "L{SSHCuphers"
       2. You use "b''.join(chr(ord(b) ^ 0x36) for b in key)" in the
          test, but string.translate in the implementation. Is that for
          performance reasons?
       3. You use "len(mod[0]().digest()))" in the test but the
          implementation uses "mod().digest_size".
       4. I've seen you say that you dislike tests that contain multiple
          assertions. It looks like the new key attribute could be tested
          separately. "self.assertEqual(key, mod.key)"
       5. The "key" variable and your new comment seem to assume that
          the test is for MD5 mac, but then loops through all
          macMap.items() which includes hmac-sha1. Does that matter?
    2. test_makeMAC
       1. "(seqid,) = struct.unpack('>L', data[:4])" I didn't
          understand this. If data is b"Hi There", why do you unpack
          the first 4 bytes as a long int? I guess seqid could be any
          long int, but using the data for this purpose confuses me.
       2. Again this test method seems to be md5 specific. What about
          sha1?
 3. [branches/use-hmac-6386/twisted/conch/ssh/transport.py](branches/use-hmac-6386/twisted/conch/ssh/transport.py)
    1. _getMAC
       1. "proscribed" > "prescribed" (I think?)
       2. "interoperability with OpenSSH (at least version 5.5p1)."
          that's quite an old version of SSH (2010) is there an open
          OpenSSH ticket that we could link to here? Is there a
          compatibility policy for conch and OpenSSH?

Your changes make sense to me, but you don't seem too confident about
this branch in comment:3, so perhaps you should wait for a review by
someone more familiar with conch. But as far as I'm concerned you can
merge after at least fixing the typos and whitespace mentioned above.

-RichardW.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

Thanks for the review!

> typo "L{SSHCuphers"

Fixed in r38405

> You use `b"".join(chr(ord(b) ^ 0x36) for b in key)` in the test, but `string.translate` in the implementation. Is that for performance reasons?

Eh, sort of.  I was thinking about performance a bit, but only key material passes through this transformation, and there's relatively little key material used (certainly compared to the amount of application data involved in any SSH connection).

I really wanted to just get rid of this code.  Notice that nothing in Conch ever uses `{in,out}MAC[{1,2}]` (sorry for that notation) anymore.  It's still computed in case some application code was relying on it for some reason, but Conch itself now computes `{in,out}MAC` with a `key` attribute and uses `hmac.HMAC` along with that `key` to compute MACs.

> You use `len(mod[0]().digest()))` in the test but the implementation uses `mod().digest_size`.

True.  They should be equal, though.  Do you think it would be better to pick one and use it?

> I've seen you say that you dislike tests that contain multiple assertions. It looks like the new key attribute could be tested separately. `self.assertEqual(key, mod.key)`

Yes, thanks.  I refactored the test in r38406.

> The `key` variable and your new comment seem to assume that the test is for MD5 mac, but then loops through all macMap.items() which includes hmac-sha1. Does that matter?

I guess it's less about MD5 and more a random coincidence about the size of the MD5 digest and the size of the key material SSH uses to initialize the HMAC algorithm.  Although it's probably not a coincidence at all, the latter was probably chosen to exactly match the former.  I expanded the comment in r38406, maybe it's less obscure now.

> `(seqid,) = struct.unpack('>L', data[:4])` I didn't understand this. If data is b"Hi There", why do you unpack the first 4 bytes as a long int? I guess seqid could be any long int, but using the data for this purpose confuses me.

Yea.  It's weird.  The API of `SSHCiphers.makeMAC` forces it.  It just reconstructs the original `data` string out of the integer and the remaining string.  The test data is from the RFC so it can't change.  `SSHCiphers.makeMAC ` is public so it can't change either.

> Again this test method seems to be md5 specific. What about sha1?

Indeed.  But the RFC only includes test vectors for MD5.  Since HMAC is digest function agnostic, if the result is correct for MD5 and the SHA1 digest function is implemented correctly, it's probably not necessary to run through any SHA1 test vectors, though it'd be nice and I'd add some more tests if there were test vectors available.

> "proscribed" > "prescribed" (I think?)

Fixed in r38407.

> "interoperability with OpenSSH (at least version 5.5p1)." that's quite an old version of SSH (2010) is there an open OpenSSH ticket that we could link to here? Is there a compatibility policy for conch and OpenSSH?

That's the version I happened to have to test against.  I don't expect it works any better with newer versions of OpenSSH, but it might be interesting to have someone test that.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

(In [38446]) Branching to 'use-hmac-6386-2'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

Latest [build results http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/use-hmac-6386-2].

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set status to `closed`|
|-|-|

(In [38616]) Merge use-hmac-6386-2

Author: exarkun
Reviewer: rwall
Fixes: [#6386](https://github.com/twisted/twisted/issues/6386)

Remove the use of PyCrypto's `Crypto.Ciphers.XOR` in the implementation
of SSH MACs in Twisted Conch and replace it with use of the stdlib `hmac`
module.
